File: go/registry_login.go
package main

import (
	"fmt"
	"log"

	"helm.sh/helm/v3/pkg/action"
	"helm.sh/helm/v3/pkg/cli"
)

type login struct {
	hostname  string
	username  string
	password  string
	certFile  string
	keyFile   string
	caFile    string
	insecure  bool
	plainHTTP bool
}

func registryLogin(logger *log.Logger, settings *cli.EnvSettings, o login) error {
	actionConfig, err := initActionConfig(settings, logger)
	if err != nil {
		return fmt.Errorf("failed to init action config: %w", err)
	}

	actionConfig.RegistryClient, err = newRegistryClientTLS(
		settings,
		logger,
		o.certFile,
		o.keyFile,
		o.caFile,
		o.insecure,
		o.plainHTTP)
	if err != nil {
		return fmt.Errorf("failed to created registry client: %w", err)
	}

	return action.NewRegistryLogin(actionConfig).Run(nil, o.hostname, o.username, o.password,
		action.WithCertFile(o.certFile),
		action.WithKeyFile(o.keyFile),
		action.WithCAFile(o.caFile),
		action.WithInsecure(o.insecure),
		action.WithPlainHTTPLogin(o.plainHTTP))
}

File: src/registry_login.rs
use thiserror::Error;

use crate::{HelmCall as _, HelmCallImpl, LoginRequest, env::Env};

#[derive(Clone, Debug, Default)]
pub struct RegistryLogin {
    pub hostname: String,
    pub username: String,
    pub password: String,
    pub cert_file: String,
    pub key_file: String,
    pub ca_file: String,
    pub insecure: bool,
    pub plain_http: bool,
    pub env: Env,
}

impl From<RegistryLogin> for LoginRequest {
    fn from(req: RegistryLogin) -> Self {
        LoginRequest {
            hostname: req.hostname,
            username: req.username,
            password: req.password,
            cert_file: req.cert_file,
            key_file: req.key_file,
            ca_file: req.ca_file,
            insecure: req.insecure,
            plain_http: req.plain_http,
            env: req.env.into(),
        }
    }
}

#[derive(Error, Debug)]
pub enum RegistryLoginError {
    #[error("registry login error: {err}")]
    RegistryLogin {
        err: String,
    },
}

pub async fn registry_login(req: RegistryLogin) -> Result<(), RegistryLoginError> {
    let res = HelmCallImpl::registry_login(req.into()).await;
    if let Some(err) = res.0.err.first() {
        return Err(RegistryLoginError::RegistryLogin {
            err: err.clone(),
        });
    }

    Ok(())
}
